Make ad_util.zero a class that carries avals (similar to UndefinedPrimal)

apaszke opened the pull request:
This is useful for remat transpose rule submitted in #3162 and e.g.
allowed me to catch a slight overuse of defjvp2 for `random_gamma_p` (it
was unnecessarily declared as having multiple outputs).

apaszke committed:
Make ad_util.zero a class that carries avals (similar to UndefinedPrimal)

This is useful for remat transpose rule submitted in #3162 and e.g.
allowed me to catch a slight overuse of defjvp2 for `random_gamma_p` (it
was unnecessarily declared as having multiple outputs).

apaszke committed:
Fix host_callback

mattjj commented on jax/interpreters/ad.py:
```diff
@@ -506,34 +509,37 @@ def defbilinear_broadcasting(bcast, prim, lhs_rule, rhs_rule):
 def instantiate_zeros(example, tangent):
-  if tangent is zero:
+  if type(tangent) is Zero:
     return zeros_like_jaxval(example)
```
We don't need to pass `example` into this function anymore AIUI, though it might be annoying to update all the call sites.

Instead, perhaps we should add a check that `example` is consistent with `tangent.aval` when `type(tangent) is Zero`? WDYT?

apaszke replied:
Oh that's really nice, I didn't realize that. Let me update the call sites!

mattjj commented on jax/interpreters/ad.py:
```diff
@@ -506,34 +509,37 @@ def defbilinear_broadcasting(bcast, prim, lhs_rule, rhs_rule):
 def instantiate_zeros_aval(aval, tangent):
-  if tangent is zero:
+  if type(tangent) is Zero:
     return zeros_like_aval(aval)
```
Similarly here, we could have a check that the avals agree (or change this function to be unary and adapt all call sites).

mattjj reviewed (approved):
LGTM, small suggestion related to additional checks we could do.

apaszke committed:
Remove example from ad.instantiate_zeros, fix vmap bug

apaszke commented:
Turns out removing the example/aval arguments was really tricky! I did manage to get rid of it for `instantiate_zeros`, which has uncovered a bug in `vmap` that I've fixed, but it turns out that `instantiate_zeros_aval` is sometimes used to replace abstract unit zeros with some shaped avals, so I had to leave it as is.

apaszke committed:
Update notebooks

apaszke committed:
Optimize zeros_like_shaped_array

This function is used a lot more now, because `ad.instantiate_zeros` now
goes through that and not `zeros_like_array`.

apaszke committed:
Remove accidental prints

mattjj reviewed (approved):
The use of `lax_reference` LGTM as well!

apaszke committed:
Rebase fixes
